retrofit-mock/src/main/java/retrofit2/mock/BehaviorDelegate.java

```diff
@@ -54,11 +60,91 @@ public T returningResponse(@Nullable Object response) {
     final Call<R> behaviorCall = new BehaviorCall<>(behavior, executor, call);
     return (T) Proxy.newProxyInstance(service.getClassLoader(), new Class[] { service },
         (proxy, method, args) -> {
-          Type returnType = method.getGenericReturnType();
+          ServiceMethodAdpaterInfo adapterInfo = parseServiceMethodAdapterInfo(method);
+
           Annotation[] methodAnnotations = method.getAnnotations();
           CallAdapter<R, T> callAdapter =
-              (CallAdapter<R, T>) retrofit.callAdapter(returnType, methodAnnotations);
-          return callAdapter.adapt(behaviorCall);
+              (CallAdapter<R, T>) retrofit.callAdapter(adapterInfo.responseType, methodAnnotations);
+
+          T adapted = callAdapter.adapt(behaviorCall);
+          if (!adapterInfo.isSuspend) {
+            return adapted;
+          }
+
+          Call<Object> adaptedCall = (Call<Object>) adapted;
+          Continuation<Object> continuation = (Continuation<Object>) args[args.length - 1];
+          try {
+            if (adapterInfo.wantsResponse) {
+              return KotlinExtensions.awaitResponse(adaptedCall, continuation);
+            }
+            return KotlinExtensions.await(adaptedCall, continuation);
+          } catch (Exception e) {
+            return KotlinExtensions.suspendAndThrow(e, continuation);
+          }
         });
   }
+
+  private static ServiceMethodAdpaterInfo parseServiceMethodAdapterInfo(Method method) {
```

Comment:
This method could be named "technical debt", but whevs. Maybe in Retrofit 3 we'll share the parsing model with the main artifact.

Reply:
Yeah. Though _that_ API gets complicated

Reply:
Mind adding this comment?

---

Computes the response type as `Call<T>` in any of these cases:

```
fun method(..., lastArg: Continuation<T>)
fun method(..., lastArg: Continuation<Response<T>>)
fun method(..., lastArg: Continuation<? extends Response<T>>)
```

Otherwise the response type is the generic return type.

Reply:
> Though that API gets complicated

I think maybe we'll just combine these artifacts so it can be internal.

Or maybe Retrofit 3 will never exist 🤞 . Or maybe it'll be a Kotlin multiplatform compiler plugin. Who knows...

> Mind adding this comment?

Done.